File: sessao-5/tutorial-19-multiarquivo-agentes/exercicios/gabarito_carrinho.ts
/**
 * Gabarito — Módulo de Carrinho de Compras com Frete (consistente)
 * Referência: Tutorial 11 — Geração multi-arquivo com agentes
 * Execute: npx ts-node gabarito_carrinho.ts
 *
 * Correção em relação a exercicio_carrinho.ts:
 *   - fecharPedido agora chama calcularTotalPedido(pedido.itens, pedido.regiao).
 *   - O frete da região é calculado e incluído no total retornado.
 *   - O dict de retorno expõe subtotal, frete e total separadamente.
 *
 * Diferença cross-file resolvida:
 *   ANTES: total = _calcularTotalLocal(pedido.itens)                    <- sem frete
 *   DEPOIS: resultado = calcularTotalPedido(itens, pedido.regiao)       <- com frete
 */

// ─── Constantes de frete (replicadas para auto-contenção) ─────────────────────

const FRETE_POR_REGIAO: Record<string, number> = {
  sudeste:    15.90,
  sul:        18.90,
  nordeste:   29.90,
  norte:      39.90,
  centroeste: 24.90,
};

const FRETE_GRATIS_ACIMA = 299.00;

// ─── Interfaces ───────────────────────────────────────────────────────────────

interface ItemPedido {
  produtoId:     string;
  descricao:     string;
  precoUnitario: number;
  quantidade:    number;
}

interface Pedido {
  id:     string;
  itens:  ItemPedido[];
  regiao: string;
}

interface ResumoPedido {
  pedidoId:  string;
  qtdItens:  number;
  regiao:    string;
  subtotal:  number;
  frete:     number;
  total:     number;
}

// ─── Lógica de precificação (alinhada com gabarito_precificacao.ts) ───────────

function _subtotalItem(item: ItemPedido): number {
  return Math.round(item.precoUnitario * item.quantidade * 100) / 100;
}

function _calcularFrete(subtotal: number, regiao: string): number {
  const regiaonorm = regiao.toLowerCase().trim();
  if (!(regiaonorm in FRETE_POR_REGIAO)) {
    const regioes = Object.keys(FRETE_POR_REGIAO).sort().join(", ");
    throw new Error(`Região desconhecida: '${regiao}'. Regiões válidas: ${regioes}`);
  }
  if (subtotal >= FRETE_GRATIS_ACIMA) return 0;
  return FRETE_POR_REGIAO[regiaonorm];
}

function calcularTotalPedido(itens: ItemPedido[], regiao: string): ResumoPedido {
  if (itens.length === 0) throw new Error("A lista de itens não pode ser vazia");
  const subtotal = itens.reduce((acc, i) => acc + _subtotalItem(i), 0);
  const frete = _calcularFrete(subtotal, regiao);
  return {
    pedidoId:  "",   // preenchido em fecharPedido
    qtdItens:  0,    // preenchido em fecharPedido
    regiao,
    subtotal:  Math.round(subtotal * 100) / 100,
    frete,
    total:     Math.round((subtotal + frete) * 100) / 100,
  };
}

// ─── Operações de pedido ──────────────────────────────────────────────────────

function criarPedido(id: string): Pedido {
  return { id, itens: [], regiao: "" };
}

function adicionarItem(pedido: Pedido, item: ItemPedido): void {
  const existente = pedido.itens.find(i => i.produtoId === item.produtoId);
  if (existente) {
    existente.quantidade += item.quantidade;
  } else {
    pedido.itens.push({ ...item });
  }
}

function definirRegiao(pedido: Pedido, regiao: string): void {
  pedido.regiao = regiao.toLowerCase().trim();
}

function fecharPedido(pedido: Pedido): ResumoPedido {
  /**
   * CORREÇÃO: chama calcularTotalPedido(pedido.itens, pedido.regiao) —
   * agora passa a região corretamente; o frete é incluído no total.
   */
  const resultado = calcularTotalPedido(pedido.itens, pedido.regiao);  // ← correto: passa regiao

  return {
    ...resultado,
    pedidoId: pedido.id,
    qtdItens: pedido.itens.reduce((acc, i) => acc + i.quantidade, 0),
  };
}

function formatarPedido(pedido: Pedido): string {
  const linhas: string[] = [`Pedido ${pedido.id}:`];
  for (const item of pedido.itens) {
    linhas.push(
      `  [${item.produtoId}] ${item.descricao}: ` +
      `R$ ${item.precoUnitario.toFixed(2)} x ${item.quantidade} = ` +
      `R$ ${_subtotalItem(item).toFixed(2)}`
    );
  }
  const subtotal = pedido.itens.reduce((acc, i) => acc + _subtotalItem(i), 0);
  linhas.push(`  Subtotal: R$ ${subtotal.toFixed(2)}`);
  if (pedido.regiao) {
    linhas.push(`  Região:   ${pedido.regiao}`);
  }
  return linhas.join("\n");
}

// ─── Execução de demonstração ─────────────────────────────────────────────────

const pedido = criarPedido("P-2026-042");

adicionarItem(pedido, { produtoId: "P001", descricao: "Teclado mecânico", precoUnitario:  89.90, quantidade: 1 });
adicionarItem(pedido, { produtoId: "P002", descricao: "Mouse sem fio",    precoUnitario:  49.90, quantidade: 2 });

definirRegiao(pedido, "nordeste");

console.log("=== Carrinho de Compras com Frete (gabarito — total com frete correto) ===\n");
console.log(formatarPedido(pedido));
console.log();

const resumo = fecharPedido(pedido);
console.log("Resumo do pedido:");
console.log(`  Pedido:   ${resumo.pedidoId}`);
console.log(`  Itens:    ${resumo.qtdItens}`);
console.log(`  Região:   ${resumo.regiao}`);
console.log(`  Subtotal: R$ ${resumo.subtotal.toFixed(2)}`);
console.log(`  Frete:    R$ ${resumo.frete.toFixed(2)}  (nordeste)`);
console.log(`  Total:    R$ ${resumo.total.toFixed(2)}  <- CORRETO: frete incluído`);

console.log();

// Pedido com frete grátis
const pedido2 = criarPedido("P-2026-043");
adicionarItem(pedido2, { produtoId: "P010", descricao: 'Monitor 4K 27"', precoUnitario: 1_299.00, quantidade: 1 });
definirRegiao(pedido2, "sul");

const resumo2 = fecharPedido(pedido2);
console.log(`Pedido ${pedido2.id} — Sul, subtotal R$ ${resumo2.subtotal.toFixed(2)}:`);
console.log(`  Frete:    R$ ${resumo2.frete.toFixed(2)}  (grátis — acima de R$ ${FRETE_GRATIS_ACIMA.toFixed(2)})`);
console.log(`  Total:    R$ ${resumo2.total.toFixed(2)}`);

console.log();
console.log("Consistência: calcularTotalPedido(pedido.itens, pedido.regiao) — assinatura alinhada.");

File: sessao-5/tutorial-19-multiarquivo-agentes/exercicios/gabarito_carrinho.py
"""
Gabarito — Módulo de Carrinho de Compras com Frete (consistente)
Referência: Tutorial 11 — Geração multi-arquivo com agentes
Execute: python3 gabarito_carrinho.py

Correção em relação a exercicio_carrinho.py:
  - fechar_pedido agora chama calcular_total_pedido(pedido.itens, pedido.regiao).
  - O frete da região é calculado e incluído no total retornado.
  - O dict de retorno expõe subtotal, frete e total separadamente.

Diferença cross-file resolvida:
  ANTES: total = _calcular_total_local(pedido.itens)                ← sem frete
  DEPOIS: resultado = calcular_total_pedido(itens, pedido.regiao)   ← com frete
"""

from dataclasses import dataclass, field


# ─── Constantes de frete (replicadas para auto-contenção) ─────────────────────

FRETE_POR_REGIAO: dict[str, float] = {
    "sudeste":    15.90,
    "sul":        18.90,
    "nordeste":   29.90,
    "norte":      39.90,
    "centroeste": 24.90,
}

FRETE_GRATIS_ACIMA: float = 299.00


# ─── Entidades ────────────────────────────────────────────────────────────────

@dataclass
class ItemPedido:
    produto_id:     str
    descricao:      str
    preco_unitario: float
    quantidade:     int


@dataclass
class ResultadoPedido:
    subtotal: float
    frete:    float
    total:    float
    regiao:   str


@dataclass
class Pedido:
    id:     str
    itens:  list[ItemPedido] = field(default_factory=list)
    regiao: str              = ""


# ─── Lógica de precificação (alinhada com gabarito_precificacao.py) ───────────

def _subtotal_item(item: ItemPedido) -> float:
    return round(item.preco_unitario * item.quantidade, 2)


def _calcular_frete(subtotal: float, regiao: str) -> float:
    """Cálculo de frete — idêntico a gabarito_precificacao.py."""
    regiao_norm = regiao.lower().strip()
    if regiao_norm not in FRETE_POR_REGIAO:
        regioes = ", ".join(sorted(FRETE_POR_REGIAO))
        raise ValueError(
            f"Região desconhecida: '{regiao}'. Regiões válidas: {regioes}"
        )
    if subtotal >= FRETE_GRATIS_ACIMA:
        return 0.0
    return FRETE_POR_REGIAO[regiao_norm]


def calcular_total_pedido(
    itens: list[ItemPedido],
    regiao: str,              # alinhado com gabarito_precificacao.py
) -> ResultadoPedido:
    """
    Calcula o total incluindo frete.
    Assinatura alinhada com gabarito_precificacao.py: ambos usam (itens, regiao).
    """
    if not itens:
        raise ValueError("A lista de itens não pode ser vazia")
    subtotal = sum(_subtotal_item(i) for i in itens)
    frete = _calcular_frete(subtotal, regiao)
    return ResultadoPedido(
        subtotal=round(subtotal, 2),
        frete=frete,
        total=round(subtotal + frete, 2),
        regiao=regiao,
    )


# ─── Operações de pedido ──────────────────────────────────────────────────────

def adicionar_item(pedido: Pedido, item: ItemPedido) -> None:
    """Adiciona item; acumula quantidade se produto_id já existir."""
    for existente in pedido.itens:
        if existente.produto_id == item.produto_id:
            existente.quantidade += item.quantidade
            return
    pedido.itens.append(item)


def definir_regiao(pedido: Pedido, regiao: str) -> None:
    """Define a região de entrega do pedido."""
    pedido.regiao = regiao.lower().strip()


def fechar_pedido(pedido: Pedido) -> dict:
    """
    Fecha o pedido e retorna um resumo com frete incluído.

    CORREÇÃO: chama calcular_total_pedido(pedido.itens, pedido.regiao) —
    agora passa a região corretamente; o frete é incluído no total.
    """
    resultado = calcular_total_pedido(pedido.itens, pedido.regiao)  # ← correto: passa regiao

    return {
        "pedido_id": pedido.id,
        "qtd_itens": sum(i.quantidade for i in pedido.itens),
        "regiao":    resultado.regiao,
        "subtotal":  resultado.subtotal,
        "frete":     resultado.frete,
        "total":     resultado.total,
    }


def formatar_pedido(pedido: Pedido) -> str:
    """Formata o estado atual do pedido para exibição."""
    linhas = [f"Pedido {pedido.id}:"]
    for item in pedido.itens:
        linhas.append(
            f"  [{item.produto_id}] {item.descricao}: "
            f"R$ {item.preco_unitario:.2f} x {item.quantidade} = "
            f"R$ {_subtotal_item(item):.2f}"
        )
    subtotal = sum(_subtotal_item(i) for i in pedido.itens)
    linhas.append(f"  Subtotal: R$ {subtotal:.2f}")
    if pedido.regiao:
        linhas.append(f"  Região:   {pedido.regiao}")
    return "\n".join(linhas)


# ─── Execução de demonstração ─────────────────────────────────────────────────

if __name__ == "__main__":
    print("=== Carrinho de Compras com Frete (gabarito — total com frete correto) ===\n")

    pedido = Pedido(id="P-2026-042")

    adicionar_item(pedido, ItemPedido("P001", "Teclado mecânico",  89.90, 1))
    adicionar_item(pedido, ItemPedido("P002", "Mouse sem fio",      49.90, 2))

    definir_regiao(pedido, "nordeste")

    print(formatar_pedido(pedido))
    print()

    resumo = fechar_pedido(pedido)
    print("Resumo do pedido:")
    print(f"  Pedido:   {resumo['pedido_id']}")
    print(f"  Itens:    {resumo['qtd_itens']}")
    print(f"  Região:   {resumo['regiao']}")
    print(f"  Subtotal: R$ {resumo['subtotal']:.2f}")
    print(f"  Frete:    R$ {resumo['frete']:.2f}  (nordeste)")
    print(f"  Total:    R$ {resumo['total']:.2f}  <- CORRETO: frete incluído")

    print()

    # Pedido com frete grátis (subtotal > FRETE_GRATIS_ACIMA)
    pedido2 = Pedido(id="P-2026-043")
    adicionar_item(pedido2, ItemPedido("P010", "Monitor 4K 27\"",    1_299.00, 1))
    definir_regiao(pedido2, "sul")

    resumo2 = fechar_pedido(pedido2)
    print(f"Pedido {pedido2.id} — Sul, subtotal R$ {resumo2['subtotal']:.2f}:")
    print(f"  Frete:    R$ {resumo2['frete']:.2f}  (grátis — acima de R$ {FRETE_GRATIS_ACIMA:.2f})")
    print(f"  Total:    R$ {resumo2['total']:.2f}")

    print()
    print("Consistência: calcular_total_pedido(pedido.itens, pedido.regiao) — assinatura alinhada.")
